src/entrypoint.cpp

```diff
@@ -271,6 +310,82 @@ namespace aperf {
+
+        SubcommandData metric_command_data = {"", "", 0, ""};
+
+        for (std::string &option : matches_metric) {
+          std::vector<std::string> parsed_option = tokenize(option, ':');
+          if (parsed_option.size() == 2) {
+            if (parsed_option[0] == "c")
+              if (metric_command_data.metric_command.empty()) {
+                metric_command_data.metric_command = parsed_option[1];
+              } else {
+                print("Cannot specify multiple metric commands", true, true);
+                return 2;
+              } 
+            else if (parsed_option[0] == "n")
+              if (metric_command_data.metric_name.empty()) {
+                metric_command_data.metric_name = parsed_option[1];
+              } else {
+                print("Cannot specify multiple metric names", true, true);
+                return 2;
+              }
+            else if (parsed_option[0] == "f")
+              if (metric_command_data.frequency == 0){
+                try
+                {
+                  metric_command_data.frequency = std::stoi(parsed_option[1]);
+                  if (metric_command_data.frequency < 0) {
+                    print("Cannot have negative frequency", true, true);
+                    return 2;
+                  }
+                } catch (const std::invalid_argument &e) {
+                  print("Error: Invalid argument. Could not convert string to integer.", true, true);
+                  return 2;
+                } catch (const std::out_of_range &e) {
+                  print("Error: Value out of range. ", true, true);
+                  return 2;
+                }
+                if (metric_command_data.frequency == 0) {
+                  print("Cannot set frequency value to 0.",true,true);
+                  return 2;
+                }
+              } else {
+                print("Cannot specify multiple frequencies for the metric", true, true);
+                return 2;
+              }
+            else if (parsed_option[0] == "r")
+              if (metric_command_data.regular_expression.empty()) {
+                metric_command_data.regular_expression = parsed_option[1];
+              } else {
+                print("Cannot specify multiple regular expressions for one metric command", true, true);
+                return 2;
+              } else {
+              print("Metric command option not recognised.", true, true);
+              return 2;
+            }
+          } else {
+            print("Cannot parse option for metric command (too many or too little agruments were given)", true, true);
+            return 2;
+          }
+        }
+        if (metric_command_data.metric_command.empty()) {
+          print("Metric command needs to be specified", true, true);
+          return 2;
+        }
+        if (metric_command_data.metric_name.empty()) {
+          print("Metric name needs to be specified", true, true);
+          return 2;
+        }
```

Comment:
Command parsing errors are handled by [CLI11 validators](https://cliutils.github.io/CLI11/book/chapters/validators.html) rather than AdaptivePerf ```print()``` functions (have a look e.g. at how event strings are validated in the code), so I would try to use these.

While implementing this suggestion, see if you can still parse user-provided metric strings only once in the code.

Reply:
Put almost everything in the check function. Added a vector to save the command arguments and iterated that afterwards to add the metric profilers in the list.